opm/autodiff/BlackoilModelBase_impl.hpp

```diff
@@ -1610,11 +1610,33 @@ namespace detail {
                 break;
 
             case SURFACE_RATE:
-                for (int phase = 0; phase < np; ++phase) {
-                    if (distr[phase] > 0.0) {
-                        xw.wellRates()[np*w + phase] = target * distr[phase];
+                // assign target value as initial guess for injectors and
+                // single phase producers (orat, grat, wrat)
+                const WellType& well_type = wells().type[w];
+                if (well_type == INJECTOR) {
+                    for (int phase = 0; phase < np; ++phase) {
+                        const double& compi = wells().comp_frac[np * w + phase];
+                        if (compi > 0.0) {
+                            xw.wellRates()[np*w + phase] = target * compi;
+                        }
+                    }
+                } else if (well_type == PRODUCER) {
+                    // for single phase producers sum of distr should be 1.0
+                    double sumdistr = distr[0];
+                    for (int phase = 1; phase < np; ++phase) {
+                        sumdistr += distr[phase];
+                    }
+                    std::cout << sumdistr << std::endl;
+                    for (int phase = 0; phase < np; ++phase) {
+                        if (distr[phase] > 0.0 && sumdistr == 1.0 ) {
```

Comment:
Except for comparing against 0.0 (or other constants you explicitly control, in some contexts), it is very fragile to compare floating point numbers like this. It is better to avoid the issue completely by checking if `distr[phase] > 0.0` for more than one phase instead of using the sum.